### PR TITLE
Add step to project generator for example code, RESTEasy is optional

### LINKED ISSUE #301: Codestarts
@fbricon asked me to start a discussion on https://quarkus.io/blog/codestarts-preview/

If vscode-quarkus is using code.quarkus.io/api, codestarts will be activated by default in the backend, so nearly transparent on your side, changes: 
- new query param: `ne (no-example): boolean`
- ignored query params: `p (path)` and `c (class name)`

Any thoughts?

### LINKED ISSUE #322: 'RESTEasy JAX-RS' should no longer be added by default
https://code.quarkus.io/ no longer adds 'RESTEasy JAX-RS' by default, so the new project wizard should do the same

### PR DESCRIPTION
**Pull request by datho7561**
Reads the data from code.quarkus.io or code.quarkus.redhat.com to see which quarkus extensions are required. This means if code.quarkus.io is used, then resteasy is not required. Reads the Open API of the code quarkus instance to determine if the option to prevent sample code being generated is present. If it is, then it adds another step to the project generation wizard to specify if sample code should be generated.

Closes #322

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

File: src/wizards/generateProject/generationWizard.ts [hunk 1]
```diff
@@ -139,7 +139,26 @@ export async function generateProjectWizard() {
       prompt: 'Your resource name',
       validate: validateResourceName
     });
-    return (input: MultiStepInput) => ExtensionsPicker.createExtensionsPicker(input, state, { showLastUsed: true, showRequiredExtensions: true, allowZeroExtensions: true });
+    return (input: MultiStepInput) => ExtensionsPicker.createExtensionsPicker(
+      input, state, { showLastUsed: true, showRequiredExtensions: true, allowZeroExtensions: true },
+      inputGenerateSampleCode);
+  }
+
+  async function inputGenerateSampleCode(input: MultiStepInput, state: Partial<ProjectGenState>) {
```
Comment (datho7561) [inline src/wizards/generateProject/generationWizard.ts:155]: I should make this step conditional on if the API supports not generating sample code by looking at the openapi. code.quarkus.io supports exclusing examples, but code.quarkus.redhat.com doesn't

File: src/definitions/QExtension.ts [hunk 2]
```diff
@@ -28,14 +28,14 @@ export class QExtension {
   artifactId: string;
   isRequired: boolean;
 
-  constructor(name: string, category: string, description: string, labels: string[], groupId: string, artifactId: string) {
+  constructor(name: string, category: string, description: string, labels: string[], groupId: string, artifactId: string, isRequired) {
```
Comment (datho7561) [inline src/definitions/QExtension.ts]: `: boolean`

File: src/definitions/QExtension.ts [hunk 3]
```diff
@@ -76,4 +76,5 @@ export interface APIExtension {
   shortName: string;
   category: string;
   order: Number;
+  default: boolean; // Uses a TypeScript builtin :|
```
Comment (datho7561) [inline src/definitions/QExtension.ts]: Remove comment

File: src/test/vscodeUiTest/ProjectGenerationWizard.ts [hunk 4]
```diff
@@ -17,6 +17,7 @@ import * as _ from 'lodash';
 
 import { InputBox, QuickPickItem, Workbench, WebDriver, WebElement, By, until, Key } from 'vscode-extension-tester';
 import { DialogHandler, OpenDialog } from 'vscode-extension-tester-native';
+import { QuickPick } from 'vscode';
```
Comment (datho7561) [inline src/test/vscodeUiTest/ProjectGenerationWizard.ts]: Remove

File: src/test/vscodeUiTest/ProjectGenerationWizard.ts [hunk 5]
```diff
@@ -213,6 +217,13 @@ export class ProjectGenerationWizard extends InputBox {
     return (await this.getNthQuickPickItemInfo(n)).label;
   }
 
+  public async highlightQuickPick(n: number): Promise<void> {
+    const theOneYouWant: QuickPickItem = await this.findQuickPick(n);
```
Comment (datho7561) [inline src/test/vscodeUiTest/ProjectGenerationWizard.ts]: better name

File: src/test/vscodeUiTest/suite/projectGenerationTest.ts [hunk 6]
```diff
@@ -100,76 +100,84 @@ describe('Project generation tests', function() {
     expect(resourceName).equals('GreetingResource');
     await wizard.next();
 
-    await wizard.sendKeys(Key.DOWN, Key.UP);
+    // await wizard.sendKeys(Key.DOWN, Key.UP);
```
Comment (datho7561) [inline src/test/vscodeUiTest/suite/projectGenerationTest.ts]: remove

File: src/utils/codeQuarkusApiUtils.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import { IncomingMessage } from "http";
```
Comment (datho7561) [inline src/utils/codeQuarkusApiUtils.ts:16]: license header

File: src/utils/codeQuarkusApiUtils.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import * as https from "https";
+import * as yaml from "js-yaml";
+import * as path from "path";
+import { QuarkusConfig } from "../QuarkusConfig";
+import { read } from "fs-extra";
+import { URL } from "url";
+
+const HTTP_MATCHER = new RegExp('^http://');
+
+export namespace CodeQuarkusApiUtils {
```
Comment (datho7561) [inline src/utils/codeQuarkusApiUtils.ts]: don't use namespace here, just export individual functions

File: src/utils/tasksUtils.ts [hunk 9]
```diff
@@ -51,11 +51,18 @@ export async function getRunningQuarkusDevTasks(workspaceFolder: WorkspaceFolder
 export function getTaskExecutionWorkingDir(task: Task): string|undefined {
   if (!task.execution) return undefined;
 
-  if (!task.execution.options || !task.execution.options.cwd) {
+  const execution: ProcessExecution | ShellExecution | CustomExecution = task.execution;
```
Comment (datho7561) [inline src/utils/tasksUtils.ts]: These changes are for removing a compilation error that's present in the master branch. Maybe they should be moved to a different PR.

**Comment by datho7561** [comment]
@fbricon did you want this PR as a part of the next vscode-quarkus release? This adds a step to the project generator to specify if example code should be generated

**Comment by fbricon** [comment]
yes add it please

**Comment by datho7561** [comment]
~~This is currently broken; the logic I used to check the openapi for the parameter to pass does not work.~~ okay should be good now

**Comment by datho7561** [comment]
I get:
Activating extension 'redhat.vscode-quarkus' failed: countries_and_timezones_1.getTimezone is not a function.
and I don't know why

File: package.json [hunk 10]
```diff
@@ -246,7 +246,6 @@
     ]
   },
   "scripts": {
-    "preinstall": "npx npm-force-resolutions",
```
Comment (rgrunber) [inline package.json]: is this meant to be removed ?

**Review by rgrunber — COMMENTED** [review]
Change looks good to me. Just one question listed inline.

File: package.json (hunk 10 above)
Comment (datho7561) [reply package.json]: I removed it because sometimes it interferes when reinstalling dependencies. I'll add it back.

**Comment by datho7561** [comment]
The tests are timing out. This is likely because this PR introduces a REST API call into the project generator to figure out the capabilities of the project generator API.

**Comment by datho7561** [comment]
I've increased the timeout time for the tests

**Review by rgrunber — APPROVED** [review]
LGTM. I think this is ready to merge.

**Comment by datho7561** [comment]
Thanks for the review, Roland!
